File: bindings/lua/satox/nft_manager.lua
--[[
Satox SDK NFT Manager Binding
]]

local nft_manager = {}

-- Private variables
local initialized = false
local callbacks = {}

-- Initialize the NFT manager
function nft_manager.initialize(config)
    config = config or {}
    
    if initialized then
        return false, "NFTManager already initialized"
    end
    
    local success, result = pcall(function()
        return _G.satox_native.nft_initialize(config)
    end)
    
    if not success then
        return false, "Failed to initialize NFTManager: " .. tostring(result)
    end
    
    initialized = result
    return result
end

-- Create an NFT
function nft_manager.create_nft(config)
    if not initialized then
        return nil, "NFTManager not initialized"
    end
    
    local success, result = pcall(function()
        return _G.satox_native.nft_create(config)
    end)
    
    if not success then
        return nil, "Failed to create NFT: " .. tostring(result)
    end
    
    return result
end

-- Get NFT info
function nft_manager.get_nft(nft_id)
    if not initialized then
        return nil, "NFTManager not initialized"
    end
    
    local success, result = pcall(function()
        return _G.satox_native.nft_get(nft_id)
    end)
    
    if not success then
        return nil, "Failed to get NFT: " .. tostring(result)
    end
    
    return result
end

-- Transfer NFT
function nft_manager.transfer_nft(nft_id, from_address, to_address)
    if not initialized then
        return nil, "NFTManager not initialized"
    end
    
    local success, result = pcall(function()
        return _G.satox_native.nft_transfer(nft_id, from_address, to_address)
    end)
    
    if not success then
        return nil, "Failed to transfer NFT: " .. tostring(result)
    end
    
    return result
end

-- Get NFTs owned by address
function nft_manager.get_nfts_by_owner(address)
    if not initialized then
        return nil, "NFTManager not initialized"
    end
    
    local success, result = pcall(function()
        return _G.satox_native.nft_get_by_owner(address)
    end)
    
    if not success then
        return nil, "Failed to get NFTs by owner: " .. tostring(result)
    end
    
    return result
end

-- Update NFT metadata
function nft_manager.update_nft_metadata(nft_id, metadata)
    if not initialized then
        return nil, "NFTManager not initialized"
    end
    
    local success, result = pcall(function()
        return _G.satox_native.nft_update_metadata(nft_id, metadata)
    end)
    
    if not success then
        return nil, "Failed to update NFT metadata: " .. tostring(result)
    end
    
    return result
end

-- Register a callback for NFT events
function nft_manager.register_callback(callback)
    if type(callback) == "function" then
        table.insert(callbacks, callback)
    end
end

-- Unregister a callback
function nft_manager.unregister_callback(callback)
    for i, cb in ipairs(callbacks) do
        if cb == callback then
            table.remove(callbacks, i)
            break
        end
    end
end

-- Trigger callbacks
function nft_manager._trigger_callbacks(event, data)
    for _, callback in ipairs(callbacks) do
        pcall(callback, event, data)
    end
end

-- Check if initialized
function nft_manager.is_initialized()
    return initialized
end

return nft_manager 

File: bindings/typescript/src/nft-manager.ts
/**
 * Satox SDK NFT Manager Binding
 */

import { SatoxError } from './error';
import { NFT } from './types';

export class NFTManager {
    private core: any;
    private initialized: boolean = false;
    private callbacks: Array<(event: string, data: any) => void> = [];

    constructor() {
        this.core = require('./native/satox_core');
    }

    async initialize(config: Record<string, any>): Promise<boolean> {
        if (this.initialized) {
            throw new SatoxError('NFTManager already initialized');
        }

        try {
            const result = await this.core.nftInitialize(config);
            this.initialized = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to initialize NFTManager: ${error.message}`);
        }
    }

    async createNFT(config: Record<string, any>): Promise<NFT> {
        this.ensureInitialized();

        try {
            return await this.core.nftCreate(config);
        } catch (error) {
            throw new SatoxError(`Failed to create NFT: ${error.message}`);
        }
    }

    async getNFT(nftId: string): Promise<NFT> {
        this.ensureInitialized();

        try {
            return await this.core.nftGet(nftId);
        } catch (error) {
            throw new SatoxError(`Failed to get NFT: ${error.message}`);
        }
    }

    async updateNFT(nftId: string, metadata: Record<string, any>): Promise<boolean> {
        this.ensureInitialized();

        try {
            return await this.core.nftUpdate(nftId, metadata);
        } catch (error) {
            throw new SatoxError(`Failed to update NFT: ${error.message}`);
        }
    }

    async transferNFT(nftId: string, fromAddress: string, toAddress: string): Promise<string> {
        this.ensureInitialized();

        try {
            return await this.core.nftTransfer(nftId, fromAddress, toAddress);
        } catch (error) {
            throw new SatoxError(`Failed to transfer NFT: ${error.message}`);
        }
    }

    async listNFTs(owner?: string): Promise<NFT[]> {
        this.ensureInitialized();

        try {
            return await this.core.nftList(owner);
        } catch (error) {
            throw new SatoxError(`Failed to list NFTs: ${error.message}`);
        }
    }

    registerCallback(callback: (event: string, data: any) => void): void {
        this.callbacks.push(callback);
    }

    unregisterCallback(callback: (event: string, data: any) => void): void {
        const index = this.callbacks.indexOf(callback);
        if (index > -1) {
            this.callbacks.splice(index, 1);
        }
    }

    private ensureInitialized(): void {
        if (!this.initialized) {
            throw new SatoxError('NFTManager not initialized');
        }
    }
} 